corehq/apps/users/tasks.py

```diff
@@ -376,6 +360,29 @@ def process_reporting_metadata_staging():
         process_reporting_metadata_staging.delay()
 
 
+def _process_reporting_metadata_staging(chunk_size=100):
+    from corehq.apps.users.models import CouchUser, UserReportingMetadataStaging
+    for pk in UserReportingMetadataStaging.objects.values_list('pk', flat=True).order_by('pk')[:chunk_size]:
+        with transaction.atomic():
+            record = UserReportingMetadataStaging.objects.select_for_update(skip_locked=True).get(pk=pk)
+            user = CouchUser.get_by_user_id(record.user_id, record.domain)
+            try:
+                record.process_record(user)
+            except ResourceConflict:
+                # if the user was updated before we were able to save the processed metadata, try again
+                user = CouchUser.get_by_user_id(record.user_id, record.domain)
+                try:
+                    record.process_record(user)
+                except Exception:
+                    notify_exception(None, "Error encountered in _process_reporting_metadata_staging")
+                else:
+                    record.delete()
+            except Exception:
+                notify_exception(None, "Error encountered in _process_reporting_metadata_staging")
+            else:
+                record.delete()
```

Comment:
What do you think of refactoring this to use `retry_on` rather than using a nested duplicate exception handling? I'm thinking of something like:
```py
from corehq.apps.users.models import CouchUser, UserReportingMetadataStaging
from dimagi.utils.retry import retry_on


def _process_reporting_metadata_staging(chunk_size=100):
    for pk in UserReportingMetadataStaging.objects.values_list('pk', flat=True).order_by('pk')[:chunk_size]:
        with transaction.atomic():
            record = UserReportingMetadataStaging.objects.select_for_update(skip_locked=True).get(pk=pk)
            try:
                _process_record_for_user(record)
            except Exception:
                notify_exception(None, "Error encountered in _process_reporting_metadata_staging")
            else:
                record.delete()


@retry_on(ResourceConflict, delays=[0])
def _process_record_for_user(record):
    user = CouchUser.get_by_user_id(record.user_id, record.domain)
    record.process_record(user)
```
The number of retries can be increased by adding more items to the `delays` list. A non-zero value will cause it to sleep that number of seconds before retrying.

Reply:
Interesting stuff! I certainly did not like the nested exception handling so this looks good. I might call it `_process_record_with_retry` instead but that is just my preference. Thanks for flagging.

Reply:
See 46e347fc2637e515291a3a2499090d03664ef78c